server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1180,151 +916,315 @@ public void snapshot(final IndexCommit snapshotIndexCommit) {
+        int latest = -1;
+        Set<String> blobKeys = blobs.keySet();
+        for (String name : blobKeys) {
+            if (name.startsWith(SNAPSHOT_INDEX_PREFIX)) {
+                try {
+                    int gen = Integer.parseInt(name.substring(SNAPSHOT_INDEX_PREFIX.length()));
+                    if (gen > latest) {
+                        latest = gen;
+                    }
+                } catch (NumberFormatException ex) {
+                    logger.warn("failed to parse index file name [{}]", name);
+                }
             }
+        }
+        if (latest >= 0) {
+            try {
+                final BlobStoreIndexShardSnapshots shardSnapshots =
+                    indexShardSnapshotsFormat.read(shardContainer, Integer.toString(latest));
+                return new Tuple<>(shardSnapshots, latest);
+            } catch (IOException e) {
+                final String file = SNAPSHOT_INDEX_PREFIX + latest;
+                logger.warn(() -> new ParameterizedMessage("failed to read index file [{}]", file), e);
+            }
+        } else if (blobKeys.isEmpty() == false) {
+            logger.warn("Could not find a readable index-N file in a non-empty shard snapshot directory [{}]", shardContainer.path());
+        }
 
-            private void checkAborted() {
-                if (snapshotStatus.isAborted()) {
-                    logger.debug("[{}] [{}] Aborted on the file [{}], exiting", shardId, snapshotId, fileName);
-                    throw new IndexShardSnapshotFailedException(shardId, "Aborted");
+        // We couldn't load the index file - falling back to loading individual snapshots
+        List<SnapshotFiles> snapshots = new ArrayList<>();
+        for (String name : blobKeys) {
+            try {
+                BlobStoreIndexShardSnapshot snapshot = null;
+                if (name.startsWith(SNAPSHOT_PREFIX)) {
+                    snapshot = indexShardSnapshotFormat.readBlob(shardContainer, name);
                 }
+                if (snapshot != null) {
+                    snapshots.add(new SnapshotFiles(snapshot.snapshot(), snapshot.indexFiles()));
+                }
+            } catch (IOException e) {
+                logger.warn(() -> new ParameterizedMessage("failed to read commit point [{}]", name), e);
             }
         }
+        return new Tuple<>(new BlobStoreIndexShardSnapshots(snapshots), latest);
     }
 
-    private static final class PartSliceStream extends SlicedInputStream {
+    /**
+     * Snapshot individual file
+     * @param fileInfo file to be snapshotted
+     */
+    private void snapshotFile(BlobStoreIndexShardSnapshot.FileInfo fileInfo, IndexId indexId, ShardId snapshotShardId,
+                              SnapshotId snapshotId, IndexShardSnapshotStatus snapshotStatus, Store store) throws IOException {
+        final BlobContainer shardContainer = shardContainer(indexId, snapshotShardId);
```

Comment:
would it make sense to pass the shardContainer directly to the method?

Reply:
The problem with that is that we still need the `shardId` for the exceptions and logging (wouldn't want to change anything about that here) so it's not really simplifying the number of inputs. This may be something to clean up more nicely in a follow up though? (could just move things around a little to be able to reuse the container all the way) :)

Reply:
Sure. Feel free to apply this in a follow up or not if this makes things more complex.